src/emplacer.rs

```diff
@@ -0,0 +1,288 @@
+use std::{collections::HashMap, path::{Path, PathBuf}, sync::Arc};
+
+use anyhow::Context;
+use sha2::{Digest, Sha256};
+use url::Url;
+
+use crate::{bindle_util::{InterestingParcel, InvoiceUnderstander, WagiHandlerInfo}, wagi_config::{HandlerConfigurationSource, WagiConfiguration}};
+
+pub struct Emplacer {
+    cache_path: PathBuf,
+    source: HandlerConfigurationSource,
+}
+
+pub struct Bits {
+    pub wasm_module: Arc<Vec<u8>>,
+    pub volume_mounts: HashMap<String, String>,
+}
+
+impl Emplacer {
+    pub async fn new(configuration: &WagiConfiguration) -> anyhow::Result<Self> {
+        Self::new_from_settings(
+            &configuration.asset_cache_dir,
+            &configuration.handlers
+        ).await
+    }
+
+    async fn new_from_settings(asset_cache_dir: &PathBuf, handlers: &HandlerConfigurationSource) -> anyhow::Result<Self> {
+        let cache_path = asset_cache_dir.clone();
+        tokio::fs::create_dir_all(&cache_path).await
+            .with_context(|| format!("Can't create asset cache directory {}", cache_path.display()))?;
+        Ok(Self {
+            cache_path,
+            source: handlers.clone(),
+        })
+    }
+
+    pub async fn emplace_all(&self) -> anyhow::Result<()> {
+        match &self.source {
+            HandlerConfigurationSource::ModuleConfigFile(_) => Ok(()),
+            HandlerConfigurationSource::StandaloneBindle(bindle_base_dir, id) =>
+                self.emplace_standalone_bindle(bindle_base_dir, id).await,
+            HandlerConfigurationSource::RemoteBindle(bindle_base_url, id) =>
+                self.emplace_remote_bindle(&bindle_base_url, id).await,
+        }.with_context(|| "Error caching assets from bindle")
+    }
+
+    // TODO: NO! NO! NO!
+    pub async fn get_bits_for(&self, handler: &WagiHandlerInfo) -> anyhow::Result<Bits> {
+        let module_parcel_path = self.module_parcel_path(&handler.parcel);
+        let wasm_module = tokio::fs::read(&module_parcel_path).await
+            .with_context(|| format!("Error reading module {} from cache path {}", handler.parcel.label.name, module_parcel_path.display()))?;
+        let volume_mounts = self.asset_dir_volume_mount(&handler.invoice_id);
+        Ok(Bits {
+            wasm_module: Arc::new(wasm_module),
+            volume_mounts,
+        })
+    }
+
+    // TODO: do not like having bindle specifics here
+    pub async fn read_invoice(&self, invoice_id: &bindle::Id) -> anyhow::Result<bindle::Invoice> {
```

Comment:
If we do have a `bindle_utils` module, would this be better placed there?

Reply:
No, this refers specifically to reading through the emplacer.